### ISSUE TITLE
AddOns

### ISSUE THREAD

**Issue by sathish-a** [google/objax#41]
Hi team, First, I'm excited about this project. I was recently playing with Objax and have implemented the learning rate finder and cyclic learning rate in my pet project. Just to know if we are looking to add it as a part of the library or just an example would be sufficient? And let me know if I can contribute anything so that I can work on it?

**Comment by AlexeyKurakin** [comment]
Thanks for the interest and willingness to contribute.

This sounds like fall into category of examples. Maybe into `examples/learning_rate` subdirectory.
We haven't thought about specific location for add-ons, but may add it in the future if there is interest.

Also, do you have any feedback regarding the current structure of `examples/`?

**Comment by sathish-a** [comment]
Thanks for the suggestion @AlexeyKurakin. How about under `examples/super_convergence` ? We can put the learning rate finder, cyclic learning rate, cyclic momentum, etc. under the above-mentioned directory. I'm currently involved in modularizing the functions. Any suggestions you would like to give?

**Comment by AlexeyKurakin** [comment]
I think `examples/super_convergence` is fine. We're going to revamp structure of examples directory at some point, but then we will move your example as well.

Please add README file to the code. Other than that I don't have specific suggestions, after you start pull request we can go from there.

**Comment by AlexeyKurakin** [comment]
@sathish-a I assigned this issue to you, since you're working on adding super_convergence example

**Comment by AlexeyKurakin** [comment]
@sathish-a friendly ping, just wanted to double check if you still working on it?

**Comment by sathish-a** [comment]
Hi @AlexeyKurakin yes. It's not working as expected when there is an ema in the training op. So debugging on that line. I'll submit PR shortly.

**Comment by david-berthelot** [comment]
By the way, don't hesitate to submit a draft pull request to get feedback/help.

**Comment by sathish-a** [comment]
I'm almost done with this. I have modified the existing `optimizers.Momentum` module to accept the momentum value during the training in order to support cyclical momentum. Now everything is working as expected. Can I share the colab link here so that you guys can review it?

**Comment by david-berthelot** [comment]
Sure, you can even share a pull request. Note, tomorrow we're planning to release 1.1.0, is there any chance you could have a PR before it? If not, that's not the end the world, it'll be in next release (probably November).

**Comment by sathish-a** [comment]
Sorry @david-berthelot got held up with some personal work. Will send a colab link I think I've to modify the design of my current implementation. Lots of code redundancy. I - require some help on it. Once I get your feedback, I'll submit a draft PR. We can continue from there. What do you say?

**Comment by david-berthelot** [comment]
Sounds good.

**Comment by sathish-a** [comment]
Hi @david-berthelot, here is the [colab link](https://colab.research.google.com/drive/1iz0Wb2GtECxddP8Jl1FqaGBfK5Zibam8?usp=sharing) to the notebook. Please add the comments in the notebook itself.

**Comment by david-berthelot** [comment]
Sorry it's hard to do code suggestions in a colab, so I'll do it here.

For the updated `Momentum` class, I would suggest to allow overriding the default value in the `__call__` method rather than redefining it, it's more consistent with the rest of Objax' design.
```python
class Momentum(Module):
    def __call__(self, lr: float, grads: List[jn.ndarray], momentum: Optional[float]):
        """Updates variables and other state based on momentum SGD.
        Args:
           lr: the learning rate.
           grads: the gradients to apply.
           momentum: optional, override the default momentum.
        """
        assert len(grads) == len(self.train_vars), 'Expecting as many gradients as trainable variables'
       momentum = momentum or self.momentum 
        for g, p, m in zip(grads, self.train_vars, self.m):
            m.value = g + self.momentum * m.value
            p.value -= lr * m.value

A few random notes.

When calling Jit on a model, there's no need to pass the model variables (you can but if you don't Jit takes them from the argument automatically).
```python
# Instead of
predict = objax.Jit(model, model.vars())
# You can simply do
predict = objax.Jit(model)
```

You can save the model weights without using `deepcopy`. Obviously you can if your prefer to, but here is a simple way using `VarCollection`:
```python
model = Model(...)  # Create the model.
trained_weights = []
original_weights = model.vars().tensors()
# Now you can train the model
# and you can even save the trained weights if you want to use them later.
trained_weights.append(models.vars().tensors())

# To reinitialize to the original weights
model.vars().assign(original_weights)
# You can train the model again from the original weights.
# and so on.
```

Apart from that, let me know what kind of feedback you're expecting from me. Your colab notebook is quite big and specific feedback requests would help prioritize. Thanks!

**Comment by sathish-a** [comment]
@david-berthelot, Thanks for the feedback. I'll incorporate these changes. The reason for the big notebook is that I wanted to showcase the results of these experiments at each phase. And for the specific feedback, let's take a function `find_lr` the inputs, and the model is tightly coupled inside the function. What I imagined was like if some user wants to experiment with their own set of models and inputs they could simply pass the necessary arguments (parametrizing the models, inputs, and loss function) like a separate module or as a callback function or shall we go with the current implementation? I'm quite unsure of what would be useful for the community.

**Comment by david-berthelot** [comment]
I think everything in your notebook could be useful to the community:

1. The modified Momentum method that lets override the momentum would be useful in the objax library itself. (Actually upgrading Adam following the same principle would be useful too).
2. Finding the learning rate would be useful but probably more as a tutorial / notebook.

@AlexeyKurakin Do you want to add anything?

**Comment by sathish-a** [comment]
Alright, so first can I send an individual PR for the modifications in Momentum and Adam and then followed by the notebook inside `examples/super_convergence`?

**Comment by david-berthelot** [comment]
Yes, sounds like a plan.

**Comment by sathish-a** [comment]
@david-berthelot, wrt Adam, do I need to add functionality to override both `beta1` and `beta2` parameters or just `beta1`?

**Comment by david-berthelot** [comment]
Let's do both.
